### PR TITLE
Feature/actionable notifications ios10

### PR DESCRIPTION
**Pull request by sofokli1**
Quick way to test new functionality please remove `private` from `AppDelegate line 21` and add the following snippet to `TimelineViewModel line 84`
`let notificationDate = Date().addingTimeInterval(5)
let notificationService = ((UIApplication.shared.delegate as? AppDelegate)?.notificationService)!
notificationService.scheduleNotification(date: notificationDate, title: "Howdy!", message: "What you’re doing right now?")`

### CODE REVIEW

File: teferi.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1153,12 +1281,12 @@
...
 				CLANG_ALLOW_NON_MODULAR_INCLUDES_IN_FRAMEWORK_MODULES = NO;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
-				DEVELOPMENT_TEAM = B227VTMZ94;
+				DEVELOPMENT_TEAM = QK3BJ9S38G;
 				FRAMEWORK_SEARCH_PATHS = "$(inherited)";
 				INFOPLIST_FILE = teferi/Info.plist;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
 				OTHER_SWIFT_FLAGS = "$(inherited) \"-D\" \"COCOAPODS\" -DDEBUG";
-				PRODUCT_BUNDLE_IDENTIFIER = com.toggl.superday;
+				PRODUCT_BUNDLE_IDENTIFIER = com.toggl.superday.sofo;
```
Comment (amulware) [inline teferi.xcodeproj/project.pbxproj]: There are a bunch of changes to the development team, and bundle identifier all through this file. I'm not 100% sure this will harm us, but it seems risky.

File: teferi/AppDelegate.swift [hunk 2]
```diff
@@ -66,6 +67,10 @@ class AppDelegate : UIResponder, UIApplicationDelegate
             return true
         }
         
+        if #available(iOS 10.0, *) {
+            (self.notificationService as? PostiOSTenNotificationService)?.setUserNotificationActions()
+        }
+
```
Comment (amulware) [inline teferi/AppDelegate.swift]: This is only run when the app starts in the foreground, and will not be run if the app starts in the background, or gets activated by the user after it started in the background. Should that code be here? Does it not need to cover the other cases?

File: teferi/Models/Category.swift [hunk 3]
```diff
@@ -27,17 +27,37 @@ enum Category : String
...
         case .commute:
-            return UIColor(hexString: "#63D5EE")
+            return UIColor(hexString: self.colorHex)
         case .food:
-            return UIColor(hexString: "#FF6453")
+            return UIColor(hexString: self.colorHex)
         case .unknown:
-            return UIColor(hexString: "#CECDCD")
+            return UIColor(hexString: self.colorHex)
+        }
```
Comment (amulware) [inline teferi/Models/Category.swift]: This switch-case is a bit silly now, isn't it? :D

File: teferi/Services/CoreDataPersistencyService.swift [hunk 4]
```diff
@@ -46,20 +46,24 @@ class CoreDataPersistencyService : PersistencyService
...
+            fetchRequest.fetchLimit = numberOfTimeSlots
+        }
+        fetchRequest.sortDescriptors = [NSSortDescriptor(key: "startTime", ascending: false)]
         
         do
         {
             let results = try getManagedObjectContext().fetch(fetchRequest) as! [NSManagedObject]
             
-            let timeSlots = results.map(mapManagedObjectToTimeSlot)
+            let timeSlots = results.reversed().map(mapManagedObjectToTimeSlot)
```
Comment (amulware) [inline teferi/Services/CoreDataPersistencyService.swift]: Why is this reversing the results? Won't this invert the order of the timeline?

File: teferi/Services/CoreDataPersistencyService.swift [hunk 5]
```diff
@@ -99,6 +103,7 @@ class CoreDataPersistencyService : PersistencyService
             managedTimeSlot.setValue(category.rawValue, forKey: "category")
             
             try managedContext.save()
+            callbacks.forEach { callback in callback(timeSlot, .update) }
```
Comment (amulware) [inline teferi/Services/CoreDataPersistencyService.swift]: What is this change necessary for?

File: teferi/Services/PostiOSTenNotificationService.swift [hunk 6]
```diff
@@ -44,4 +63,58 @@ class PostiOSTenNotificationService : NotificationService
         
         notifications.forEach { n in UIApplication.shared.cancelLocalNotification(n)  }
     }
+    
+    func handleNotificationAction(withIdentifier identifier: String?, for notification: UILocalNotification, completionHandler: @escaping () -> Void) {
+        if let identifier = identifier {
+            
+            var category = Category.commute
+            
+            switch identifier {
```
Comment (amulware) [inline teferi/Services/PostiOSTenNotificationService.swift]: Rather than this conversion using constants, can't we just use the .rawValue of the category enum?

File: teferi/ViewControllers/TimelineViewController.swift [hunk 7]
```diff
@@ -67,19 +67,40 @@ class TimelineViewController : UITableViewController
...
+        case .none:
+            
+            self.tableView.reloadData()
+            
+            let scrollIndexPath = IndexPath(row: viewModel.timeSlots.count, section: 0)
+            self.tableView.scrollToRow(at: scrollIndexPath, at: .bottom, animated: true)
+            
+        default:
+            break
+        }
```
Comment (amulware) [inline teferi/ViewControllers/TimelineViewController.swift]: These are big changes, can you give a quick summary of what exactly they do?

File: teferi/ViewModels/TimelineViewModel.swift [hunk 8]
```diff
@@ -41,31 +49,48 @@ class TimelineViewModel
...
-        //Finishes last task, if needed
-        if let lastTimeSlot = timeSlots.last
-        {
-            lastTimeSlot.endTime = Date()
+        switch changeType {
+        case .create:
+            
+            if let lastTimeSlot = timeSlots.last
+            {
+                lastTimeSlot.endTime = Date()
```
Comment (amulware) [inline teferi/ViewModels/TimelineViewModel.swift]: This looks like a hack. Not important for this PR, but should be looked at. @willsb

File: teferi/ViewModels/TimelineViewModel.swift [hunk 9]
```diff
@@ -41,31 +49,48 @@ class TimelineViewModel
         self.date = date.ignoreTimeComponents()
         self.persistencyService = persistencyService
         self.timeSlotsVariable = Variable(timeSlotsForDate)
+        self.timeSlotChangeVariable = Variable(TimeSlotChangeType.none)
         self.isEditingObservable = self.isEditingVariable.asObservable()
         self.timeSlotsObservable = self.timeSlotsVariable.asObservable()
+        self.timeSlotChangeObservable = self.timeSlotChangeVariable.asObservable()
```
Comment (amulware) [inline teferi/ViewModels/TimelineViewModel.swift]: What are the new Variable and Observable used for?

File: NotificationContentExtension/NotificationViewController.swift [hunk 10]
```diff
@@ -0,0 +1,51 @@
+//
+//  NotificationViewController.swift
+//  NotificationContentExtension
+//
+//  Created by Juxhin Bakalli on 08/11/2016.
+//  Copyright © 2016 Toggl. All rights reserved.
+//
```
Comment (willsb) [inline NotificationContentExtension/NotificationViewController.swift]: //TODO: Add this to the Readme.md guidelines

File: teferi.xcodeproj/project.pbxproj (hunk 1 above)
Comment (willsb) [inline teferi.xcodeproj/project.pbxproj]: These changes should not be commited, as he just used them because of the Provisioning Profiles

Comment (sofokli1) [inline teferi.xcodeproj/project.pbxproj]: Will fix. I have worked with `fastlane` tool `match` with one shared account between developers. It did wonders regarding this kind of issues + only one user needed to know the account password (for setup reasons). It abstracts the signing and certificate creation process [https://codesigning.guide](url) Just a thought.

File: teferi/Services/CoreDataPersistencyService.swift (hunk 4 above)
Comment (sofokli1) [inline teferi/Services/CoreDataPersistencyService.swift]: Before modifying this function the order of the data was ascending. I modified the function so that it loads a specific number of timeSlots (or all) for a certain day. Since it requires to return the last N timeSlots I added a sortDescriptor to the fetchRequest in order to return the data in descending order in relation to the startTime (so that we get the latest data first). Afterwords the result from the fetchRequest is reversed so that the function returns data in ascending order as it was before modification it and as it is needed for the actionable notifications. We could add a `ascending: Bool` as a parameter to the function so that is more clear how the data are returned. What do you think @amulware ?

File: teferi/AppDelegate.swift (hunk 2 above)
Comment (sofokli1) [inline teferi/AppDelegate.swift]: Based on the documentations there isn't much need to call this function again somewhere else.

> This method registers all of your categories at once, replacing any previously registered categories with the new ones in the categories parameter. Typically, you call this method only once at launch time.

File: teferi/Services/CoreDataPersistencyService.swift (hunk 5 above)
Comment (sofokli1) [inline teferi/Services/CoreDataPersistencyService.swift]: It wasn't clear what kind of changes were happening to timeSlot so the `TimelineViewModel` was handling all of them as new timeSlot. This was needed because the changes that were done on the timeSlot based on the notification action were not propagated to the UI.

File: teferi/Services/PostiOSTenNotificationService.swift (hunk 6 above)
Comment (sofokli1) [inline teferi/Services/PostiOSTenNotificationService.swift]: That sound better. Will fix

File: teferi/ViewControllers/TimelineViewController.swift (hunk 7 above)
Comment (sofokli1) [inline teferi/ViewControllers/TimelineViewController.swift]: The function was only supporting the case were new timeSlot where added to the `[TimeSlots]`. When notification changed the last time slot it could perform the needed UI updates. So instead of monitoring the array it observes for `TimeSlotChangeType` so that it is clear what happened to the array and what kind of action it needs to perform.

File: teferi/ViewModels/TimelineViewModel.swift (hunk 9 above)
Comment (sofokli1) [inline teferi/ViewModels/TimelineViewModel.swift]: Shows the kind of changes that have happened to the `timeSlots'

File: teferi/Services/CoreDataPersistencyService.swift (hunk 4 above)
Comment (amulware) [inline teferi/Services/CoreDataPersistencyService.swift]: No need to add a parameter we don't need right now.

File: teferi/AppDelegate.swift (hunk 2 above)
Comment (sofokli1) [inline teferi/AppDelegate.swift]: For the app to run in the background or any other way it needs to at least have been opened once in the foreground, which is enough for the action registration to take place. Even if the user starts the app for the first time and does not allow notifications, the notification action will  still register. I can't think of a situation where the registration does not happen at least once (which is all we need).

**Comment by amulware** [comment]
@willsb I'll let you have the final word on this, since it's your branch and you started work on this originally.
